### DISCUSSION TITLE
`window-step-resize` using wrong step sizes on macOS

### DISCUSSION THREAD

**Discussion by roblillack** [ghostty-org/ghostty#4162] [Feature Requests, Ideas]
Step resizing is broken for me as it feels like the size increments are out of sync with the font size, leading to a seeminly random padding on the bottom and the right side of the window.

Here's a video highlighting the behavior:

https://github.com/user-attachments/assets/b56fdf93-cedd-47f6-87b3-5dd3e597894f

And another one with the inspector open:

https://github.com/user-attachments/assets/1a963281-573b-439a-8395-d965680da755

I'm on a MacBook Pro 14" and seeing this on the internal screen with "standard scaling" and my font of choice is JetBrains Mono.

These are the things that don't make a difference:
- disabling the window padding
- changing the font
- changing the font size
- changing the system's display scale factor (even completely disabling retina mode by using the displays native resolution)
 
The resize increments always seem to be way off compared to the font size (ie. font size: 26px --> vertical step size: 18px, font size: 24px --> vertical step size: 16px).

**Comment by roblillack** [comment]
Playing around with this a bit more, it seems like _some_ font family/size combinations seem to work correctly, though:
- Courier New/13pt works, with 12.5pt it seem to be broken only vertically and with 12pt it is broken horizontally as well
- Menlo/12.5pt works, Menlo/13pt is broken
- PT Mono 13pt works, PT Mono/12.5pt is broken

Additionally, in cases where it _does_ work, the initial window size sometimes is off even if the increments are correct, when a window padding != 0 is specified:

https://github.com/user-attachments/assets/ee6cfbc0-57bf-44f6-90c8-0c80f88e2897

**Comment by JakeWharton** [comment]
Same as https://github.com/ghostty-org/ghostty/discussions/2946? I feel your pain!

**Comment by roblillack** [comment]
Yes, I think #2946 and the last part of the reported issue (wrong _initial_ window size) look to be the same as #2664. Regarding the broken, font dependent resize increments. 🤷

**Comment by ilyagr** [comment]
I'm experiencing something similar. It is somewhat reminiscent of https://github.com/ghostty-org/ghostty/issues/2664, but not identical. I tested this on Ghostty 1.0.0 on Mac OS Sonoma on ARM, with the config consisting just of

```
window-step-resize=true
```

On start, the window is not perfectly aligned, but OK:

![SCR-20241228-uqkw](https://github.com/user-attachments/assets/51625e5e-33ab-4948-9eff-a2afa223987f)


One problem happens on resizing:

![SCR-20241229-badc-3](https://github.com/user-attachments/assets/20045728-9710-4342-b38a-58e52b312251)

Video:

https://github.com/user-attachments/assets/831b9040-f77a-4fa5-b2a8-ba2c2ebf1203

Another problem is that resizing with `Cmd-=` keeps the window size the same, so even if it was divisible by the size of a character cell before, it isn't after.

In #2664 (I think) someone mentioned that this could be related to High-DPI displays, and indeed I have somewhat unusual monitor setup. I'm using these [BetterDisplay](https://github.com/waydabber/BetterDisplay#readme) settings on an external 1440x2560 display (rotated vertically, in extended desktop mode). Note the HiDPI.

<img width="286" alt="SCR-20241229-bjld" src="https://github.com/user-attachments/assets/14987415-f189-4f60-9ae7-21681fd4b2c4" />

P.S. This is not a huge problem for me, I switched to using `window-padding-color=extend-always` and `window-step-resize=false`, which works out OK for me. The first of these is a great idea, haven't seen it in other terminals.

**Comment by mitchellh** [comment]
I think this discussion may be exactly this comment: https://github.com/ghostty-org/ghostty/discussions/2946#discussioncomment-11566026

Certain font sizes force cell sizes that could result in fractional size increments which macOS doesn't support, perhaps. I haven't verified this, but it sounds plausible.
